components/scim/org.wso2.carbon.identity.scim.provider/src/main/java/org/wso2/carbon/identity/scim/provider/impl/SCIMUserManager.java

```diff
@@ -698,6 +714,20 @@ public Group getGroup(String id) throws CharonException {
     @Override
     public Group updateGroup(Group oldGroup, Group newGroup) throws CharonException {
 
+        oldGroup.setDisplayName(UserCoreUtil
+                                        .addDomainToName(UserCoreUtil.removeDomainFromName(oldGroup.getDisplayName()),
+                                                         UserCoreUtil
+                                                                 .extractDomainFromName(oldGroup.getDisplayName())));
+        if (UserCoreUtil.extractDomainFromName(newGroup.getDisplayName())
+                        .equals(UserCoreConstants.PRIMARY_DEFAULT_DOMAIN_NAME) &&
+            !(UserCoreUtil.extractDomainFromName(oldGroup.getDisplayName())
+                          .equals(UserCoreConstants.PRIMARY_DEFAULT_DOMAIN_NAME))) {
+            String userStoreDomain = UserCoreUtil.extractDomainFromName(oldGroup.getDisplayName());
+            newGroup.setDisplayName(UserCoreUtil.addDomainToName(newGroup.getDisplayName(), userStoreDomain));
+            if (newGroup.getMembers() != null && !(newGroup.getMembers().isEmpty())) {
+                newGroup = addDomainToUserMembers(newGroup, userStoreDomain);
```

Comment:
The problem I see is unnecessary looping of usernames. If the looping is needed just because you need to append the domain part to the username in the request because the existing users from user store are retrieved with domain name, then I think we need to look at if we can remove the domain name when fetching the existing users from the user store. Because its quite obvious the domain of the users is same as the domain of the role itself, because we can't assign users to role cross-domain.